lightning/src/ln/channelmanager.rs

```diff
@@ -71,6 +71,181 @@ use core::time::Duration;
 use std::time::Instant;
 use core::ops::Deref;
 
+mod inbound_payment {
+	use bitcoin::hashes::{Hash, HashEngine};
+	use bitcoin::hashes::cmp::fixed_time_eq;
+	use bitcoin::hashes::hmac::{Hmac, HmacEngine};
+	use bitcoin::hashes::sha256::Hash as Sha256;
+	use chain::keysinterface::{Sign, KeysInterface};
+	use ln::{PaymentHash, PaymentPreimage, PaymentSecret};
+	use ln::channelmanager::ExpandedPmtKey;
+	use ln::msgs;
+	use ln::msgs::MAX_VALUE_MSAT;
+	use util::chacha20::ChaCha20;
+	use util::errors::APIError;
+	use util::logger::Logger;
+
+	use core::convert::TryInto;
+	use core::ops::Deref;
+
+	const IV_LEN: usize = 16;
+	const METADATA_LEN: usize = 16;
+	const METADATA_KEY_LEN: usize = 32;
+
+	pub(super) fn create<Signer: Sign, K: Deref>(keys: &ExpandedPmtKey, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32, keys_manager: &K, highest_seen_timestamp: u64) -> Result<(PaymentHash, PaymentSecret), APIError>
+		where K::Target: KeysInterface<Signer = Signer>
+	{
+		let metadata_bytes = construct_metadata_bytes(min_value_msat, false, invoice_expiry_delta_secs, highest_seen_timestamp)?;
+
+		let mut iv_bytes = [0 as u8; IV_LEN];
+		let rand_bytes = keys_manager.get_secure_random_bytes();
+		iv_bytes.copy_from_slice(&rand_bytes[..IV_LEN]);
+
+		let mut hmac = HmacEngine::<Sha256>::new(&keys.ldk_pmt_hash_key);
+		hmac.input(&iv_bytes);
+		hmac.input(&metadata_bytes);
+		let payment_preimage_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let ldk_pmt_hash = PaymentHash(Sha256::hash(&payment_preimage_bytes).into_inner());
+		let payment_secret = construct_payment_secret(&iv_bytes, &metadata_bytes, &keys.metadata_key);
+		Ok((ldk_pmt_hash, payment_secret))
+	}
+
+	pub(super) fn create_from_hash(keys: &ExpandedPmtKey, min_value_msat: Option<u64>, payment_hash: PaymentHash, invoice_expiry_delta_secs: u32, highest_seen_timestamp: u64) -> Result<PaymentSecret, APIError> {
+		let metadata_bytes = construct_metadata_bytes(min_value_msat, true, invoice_expiry_delta_secs, highest_seen_timestamp)?;
+
+		let mut hmac = HmacEngine::<Sha256>::new(&keys.user_pmt_hash_key);
+		hmac.input(&metadata_bytes);
+		hmac.input(&payment_hash.0);
+		let hmac_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let mut iv_bytes = [0 as u8; IV_LEN];
+		iv_bytes.copy_from_slice(&hmac_bytes[..IV_LEN]);
+
+		Ok(construct_payment_secret(&iv_bytes, &metadata_bytes, &keys.metadata_key))
+	}
+
+	fn construct_metadata_bytes(min_value_msat: Option<u64>, user_pmt_hash: bool, invoice_expiry_delta_secs: u32, highest_seen_timestamp: u64) -> Result<[u8; 16], APIError> {
```

Comment:
Consider making a payment method enum instead of using a bool. Would improve readability and be useful if other methods are added in the future.

Reply:
Implemented in 82e6c3a8ae44fdd0e296d172770535c8346ad784. I think it's more readable!